In [ ]:
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

In [14]:
from huggingface_hub import notebook_login
notebook_login()

In [56]:
#need to train model
#validate modela
#find embeddings for each model
from transformers import AutoTokenizer, AutoModelForCausalLM

def generate_embeddings(data, model_name):
  tokenizer = AutoTokenizer.from_pretrained(model_name)
  model = AutoModelForCausalLM.from_pretrained(model_name)

  embeddings_list = []
  prompt = 'I want you to think hard about the truth of the following sentence: '

  for item in data:
    input_text = f'{prompt} {item.rstrip('.')}'
    tokens = tokenizer(input_text, return_tensors='pt')
    output = model.generate(tokens.input_ids, attention_mask=tokens.attention_mask, output_hidden_states=True, return_dict_in_generate=True, max_new_tokens=1, min_new_tokens=1)
    states = output.hidden_states
    # num_layers = len(states[0])
    # num_layers_to_use = 3
    # layers_to_use = list(range(num_layers - num_layers_to_use, num_layers))
    # embeddings_list = []
    # for layer in layers_to_use:
      # embeddings_list.append(states[0][layer][0][-1])
    embeddings_list.append(states[0][-1][0][-1]); #last layer of every model output

  return embeddings_list

In [21]:
import pandas as pd
data = pd.read_csv('./data/animals_true_false.csv')
sentences = data['statement']
data

statement  label
0       The giant anteater uses walking for locomotion.      1
1                The eagle has a habitat of urban/wild.      0
2     The tortoise has an iridescent tail with eye-l...      0
3     Human uses for hyena include conservation, res...      0
4            The platypus uses swimming for locomotion.      1
...                                                 ...    ...
1003        The pigeon has the atomic number of mammal.      0
1004                          The kangaroo is a mammal.      1
1005                The raccoon has a diet of omnivore.      1
1006          The chimpanzee has a habitat of mountain.      0
1007                            The alpaca is a mammal.      1

[1008 rows x 2 columns]

In [ ]:
#need to read in a csv of data
model_name = "openai-community/gpt2-large"
embeddings = generate_embeddings(sentences, model_name)

In [ ]:
# Process to save model locally 
# model_path = "./llama-model"
# tokenizer.save_pretrained(model_path)
# model.save_pretrained(model_path)
embeddings

In [58]:
from sklearn.model_selection import train_test_split
import numpy as np

X = np.array(embeddings)
np.save(f'./data/{model_name.split('/')[1]}_embeddings_prompt.npy', X)

y = np.array(data['label'])
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [59]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

# Assuming your embeddings have a dimension of 768 (adjust as needed)
input_dim = 1024

model = Sequential()
model.add(Dense(128, input_dim=input_dim, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(1, activation='sigmoid'))  # Binary classification example
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

/Users/ayushjain/Development/Personal-Projects/Probing/.venv/lib/python3.12/site-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [60]:
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

In [61]:
#now need to do prediction and validation
test_loss, test_acc = model.evaluate(X_test, y_test)

In [62]:
display(test_acc)

0.5742574334144592

In [63]:
display(test_loss)

0.5972087979316711

## Data
1. GPT2 Sentences without Prompt: loss : 0.61538, acc: 0.564
2. GPT2 Sentences with Prompt: loss: .62908, acc: 0.5792
3. GPT2-Medium w/ Prompt: loss: .612, acc: .549
4. GPT2-Medium w/o Prompt: loss: .597, acc: .574
